File: Template method/python/TemplateMethod.py
import abc


class Transmitter(metaclass=abc.ABCMeta):

    def _voice_record(self):
        print("recording a fragment of speech")

    def _simpling(self):
        pass

    def _digitization(self):
        pass

    @abc.abstractmethod
    def _modulation(self):
        pass

    def _transmission(self):
        print("ignal transmission via radio channel")

    def process_start(self):
        self._voice_record()
        self._simpling()
        self._digitization()
        self._modulation()
        self._transmission()


class AnalogTransmitter(Transmitter):

    def _modulation(self):
        print("modulation of the analog signal")


class DigitTransmitter(Transmitter):

    def _simpling(self):
        print("sampling of the recorded fragment")

    def _digitization(self):
        print("digitization")

    def _modulation(self):
        print("digital signal modulation")


if __name__ == '__main__':
    analog_transmitter = AnalogTransmitter()
    analog_transmitter.process_start()

    print()

    digit_transmitter = DigitTransmitter()
    digit_transmitter.process_start()

File: Template method/c++/TemplateMethod.cpp
#include <iostream>

class Transmitter
{
    protected:
        virtual void voiceRecord();
        virtual void simpling() {};
        virtual void digitization() {};
        virtual void modulation() = 0;
        virtual void transmission();

    public:
        void processStart();
        virtual ~Transmitter(){}
};

class AnalogTransmitter : public Transmitter
{
    public:
        void modulation() override;
};

class DigitTransmitter : public Transmitter
{
    public:
        void simpling() override;
        void digitization() override;
        void modulation() override;
};

int main()
{
    Transmitter* analogTransmitter = new AnalogTransmitter();
    analogTransmitter->processStart();

    std::cout << std::endl;

    Transmitter* digitTransmitter = new DigitTransmitter();
    digitTransmitter->processStart();

    delete analogTransmitter;
    delete digitTransmitter;

    return 0;
}

void Transmitter::voiceRecord()
{
    std::cout << "recording a fragment of speech\n";
}

void Transmitter::transmission()
{
    std::cout << "signal transmission via radio channel\n";
}

void Transmitter::processStart()
{
    voiceRecord();
    simpling();
    digitization();
    modulation();
    transmission();
}

void AnalogTransmitter::modulation()
{
    std::cout << "modulation of the analog signal\n";
}

void DigitTransmitter::simpling()
{
    std::cout << "sampling of the recorded fragment\n";
}

void DigitTransmitter::digitization()
{
    std::cout << "digitization\n";
}

void DigitTransmitter::modulation()
{
    std::cout << "digital signal modulation\n";
}